Lib/unittest/mock.py

```diff
@@ -261,7 +261,7 @@ def reset_mock():
     funcopy.method_calls = _CallList()
     funcopy.mock_calls = _CallList()
 
-    funcopy.return_value = mock.return_value
+    funcopy.return_value = mock._mock_return_value
```

Comment:
Thanks @infohash for the detailed analysis. Really helped with the review.

Instead of accessing `_mock_return_value` does fixing return_value getattr to take wraps into consideration a more cleaner fix? Something like below that also fixes probably other cases. I don't see any test case failures due to this change in mock's tests.

```diff
diff --git a/Lib/unittest/mock.py b/Lib/unittest/mock.py
index 93f4d9743e..29a0e37a6c 100644
--- a/Lib/unittest/mock.py
+++ b/Lib/unittest/mock.py
@@ -573,7 +573,7 @@ def __get_return_value(self):
         if self._mock_delegate is not None:
             ret = self._mock_delegate.return_value
 
-        if ret is DEFAULT:
+        if ret is DEFAULT and self._mock_wraps is None:
             ret = self._get_child_mock(
                 _new_parent=self, _new_name='()'
```

While testing I came across below scenario where if one of the mock with wraps access return_value before call then wraps doesn't work and mock is returned as per the report. With the patch it will have sentinel.DEFAULT and the wraps call would also work. There is less usage of explicitly accessing return_value attribute to compare it with something else and it's mostly setting return_value attribute but just wanted to add this case.

```python
from unittest.mock import Mock

def wrapped():
    return "wrapped"

a = Mock(wraps=wrapped)
print(a.return_value)
print(a())

a = Mock()
print(a.return_value)
```

Without patch

```
./python /tmp/bar.py
<Mock name='mock()' id='139793330596048'>
<Mock name='mock()' id='139793330596048'>
<Mock name='mock()' id='139793330596192'>
```

With patch

```
./python /tmp/bar.py
sentinel.DEFAULT
wrapped
<Mock name='mock()' id='140270447781376'>
```

Reply:
Yes, your solution is more appropriate, I have added it. While I was testing after adding your suggested change, I have also made one more change to maintain backward compatibility with this particular statement written in the [official documentation](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock):

> **If the mock has an explicit return_value set then calls are not passed to the wrapped object and the return_value is returned instead.**

The test below will pass now and I have also included this test case in the PR.

```python
from unittest import mock


def wrapped():
    return "wrapped"


x = mock.create_autospec(spec=wrapped, wraps=wrapped)
x.return_value = "test_value"

>>> x()
'wrapped'
>>> x() == "wrapped"
False
>>> x() != "test_value"
True
```